pkg/component/operator/web/v0/crawl_website.go

```diff
@@ -148,6 +148,11 @@ func (e *execution) CrawlWebsite(input *structpb.Struct) (*structpb.Struct, erro
 		r.Headers.Set("User-Agent", randomString())
 	})
 
+	// colly.Wait() does not terminate the program. So, we need a system to terminate the program when there is no collector.
+	// We use a channel to notify the main goroutine that a new page has been scraped.
+	// When there is no new page for 2 seconds, we cancel the context.
```

Comment:
Silly question: is there a way to know there aren't remaining URLs to scrape (e.g. cancelling the context after `Wait` finishes)? Even if there's a sweet spot, waiting for any amount of time is always a compromise between being inefficient (waiting when we could return) and risking data loss (there's new data but it's taking longer than the threshold).

Reply:
Originally, I expected colly should finish the job when there aren't remaining URLs.

But, I could not find a way to close colly when there is no callback jobs.
So, I decided to use this way to deal with the case that the output length hasn't achieved the max-k but there is no URLs.

Reply:
Something that isn't clear to me is what you mean by 
> does not terminate the program. So, we need a system to terminate the program when there is no collector.

I'm not familiar with `colly`, so what I'm proposing might not have sense (in which case, I'm just trying to understand the code). Colly is a popular enough project, so my intuition tells me there has to be a simpler way to finish the execution when all the websites have been crawled. Also I'm assuming that `c.Wait()` does work and that it halts the execution until all the sites have been scraped.

From the assumption I just mentioned, what I see you're trying to accomplish is waiting until the first occurrence between:
- All the sites have been scraped. `c.Wait()` will signal this.
- We've reached the 2 minute timeout. `ctx.Done()` will signal this.
- We've reached the limit of pages. `ctx.Done()` will also signal this, as we cancel the context when we find this condition.

Therefore, we need to wait for either `c.Wait()` or `ctx.Done()`. I think we can avoid timing the duration between 2 completed pages and select between `ctx.Done()` and a channel that is closed after `c.Wait()` ends:

```go
	scrapeDone := make(chan struct{})

	go func() {
		defer close(scrapeDone)
		_ = c.Visit(inputStruct.URL)
		c.Wait()
	}()

	select {
	case <- scrapeDone:
		// No more pages to scrape.
	case <-ctx.Done():
		// Timeout or page limit reached
	}
```

Reply:
> All the sites have been scraped. `c.Wait()` will signal this.

Yes! It is the problem. I expect `c.Wait()` will finish when there is no more callback (job).
However, when I did end-to-end test, I noticed it did not close this go-routine. 
If `c.Wait()` can close it, we won't need another `chan`(If I understand it correctly).

Reply:
> I noticed it did not close this go-routine.

I don't fully understand this. How were you checking the goroutine had ended? The moment you use `go func`, the main thread stops caring about the function you're calling, so it will never know when `c.Wait` finished. That's why in  my snippet I added a channel that's closed at the end of the spawned function. By listening to that channel, the main thread is signaled about the wait group being unlocked.

Reply:
Oh, I think I get the point. I will try your proposal.

Reply:
Let me know how it goes!

Reply:
It doesn't work. 😢 .
I guess the only way  to finish `c.Wait()` is `r.Abort()` or cancel the ctx.

However, we cannot abort it if the the output page count has not achieved maxK.

Reply:
🤔 That is weird, as I mentioned the documentation indicates that `Wait` should signal the end of the scraping and the project is popular enough as to discard an issue here.

I actually [modified their parallel example](https://gist.github.com/jvallesm/b542f1b377863deafeecb8b434f996ed) (I also included URL filtering if you want to check it out) and, indeed, the strategy I suggested works.

Therefore, this claim is incorrect, there must be something else preventing the collector from finishing.
>  I guess the only way to finish c.Wait() is r.Abort() or cancel the ctx.

Reply:
I am not sure if I am right. But, I imagine like the flow blow.

When `OnHTML` is triggered, it collects the to be scraped URLs in a queue.
The workers will scrape the URLs in the queue in the sequence.

And, `OnHTML` will be triggered in a big count in the beginning, which makes the the queue contains lots of URLs that need workers to process.

So, that is the reason why the workers do not close.

But, it comes to another question. Why do we need to close the worker if they haven't finished?
It is because those remaining pages aren't probably the target pages because the crawler failed to get data by the timeout setting.

So, it seems the remaining URLs are the URLs that the workers need to process that to wait for the timeout and cannot get any data.

If this assumption is correct, we will need a timer to close the worker to avoid the worker to wait for all URLs' timeout.

The [sample](https://gist.github.com/chuang8511/4f721be9b3c9de648d725162ccefffce) is just printing out the `htmlcount` & `output count`.

Reply:
But in that scenario, we're not talking about the process having finished, it's just that each new visit takes longer and we want to set a timeout because that _probably_ means these results aren't interesting. In that case, indeed `Wait()` won't work, but then we're not listening to the end of the process. Then we can wait for the total timeout or set a timeout between requests, which I don't love but I see the point about.

In your example (and here), should we cancel the context when the `MaxK` has been reached? Otherwise requests are skipped but the crawling goes on. I don't know if this is intentional.

Also, isn't it simpler to use channels here than mutexes?

Reply:
> should we cancel the context when the MaxK has been reached?

Yes, we should. When the output reaches the count the users want, we can return the result to the users. So, we should cancel the context and stop requesting and crawling. 

> Also, isn't it simpler to use channels here than mutexes?

I am not quite sure what it specifically means in which part. We do use channels to signal the stop process here. right?

Reply:
> > should we cancel the context when the MaxK has been reached?
>
> Yes, we should. When the output reaches the count the users want, we can return the result to the users. So, we should cancel the context and stop requesting and crawling.

I mentioned that because your sample didn't do that, so the code kept crawling after the max results had been collected and it ran until the timeout was reached. If you cancel the context (or close the `scrapeDone` channel), the program doesn't reach the timeout. Adding this, we can check the 3 scenarios work (in your sample, we'd need to port that to the component code):

- No more pages to scrape: `c.Wait()` returns and the goroutine closes `scrapeDone`. Program finishes before the timeout.
- Max pages scraped: `c.OnResponse` cancels the context / closes `scrapeDone`. Program finishes before the timeout.
- Max pages haven't been collected before the timeout. Context is canceled and program finishes at the timeout.

This is the modification I made:
```go
	c.OnResponse(func(r *colly.Response) {
		log.Println(r.Request.URL.String())
		mu.Lock()
		defer mu.Unlock()
		responsecount++
		fmt.Println("responsecount", responsecount)
		output = append(output, r.Request.URL.String())
		if len(output) >= 10 {
			close(scrapeDone) // cancel()
			return
		}
		fmt.Println(r.StatusCode)
	})
```

Reply:
> > Also, isn't it simpler to use channels here than mutexes?
>
> I am not quite sure what it specifically means in which part. We do use channels to signal the stop process here. right?

What I  mean is that you're defining variables outside of the goroutines and using mutexes to make sure you modify them atomically. To me a much more natural pattern is using channels to communicate the information outside of the routines. There's [this article](https://go.dev/wiki/MutexOrChannel) and also the [go proverb](https://go-proverbs.github.io/):

> [Channels orchestrate; mutexes serialize.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=4m20s).

Reply:
> What I mean is that you're defining variables outside of the goroutines and using mutexes to make sure you modify them atomically.

I got what you mean here. You mean the output pages. right? 
I will address this point in other PR and take care of this part in the future development.


> Adding this, we can check the 3 scenarios work (in your sample, we'd need to port that to the component code):

Let's back to the current code(PR) not sample one. 
How about I do the following thing.

I will add back the `scrapeDone` channel and remain the "2 second part" in the codebase. 
So, we can deal with 4 scenarios.

1. No more pages to scrape: c.Wait() returns and the goroutine closes scrapeDone. Program finishes before the timeout.
2. Max pages scraped: c.OnResponse cancels the context / closes scrapeDone. Program finishes before the timeout.
3. Max pages haven't been collected before the timeout. Context is canceled and program finishes at the timeout.
4. Max pages have not been collected before the timeout. Context is canceled and the program finishes when there are no more validate data in 2 seconds.

And, why we need the 4th one is because what I mentioned [here](https://github.com/instill-ai/pipeline-backend/pull/819#discussion_r1843668582).

Reply:
Agreed, make sure you document the 4th in the code 🤝 I [updated the sample](https://gist.github.com/jvallesm/b542f1b377863deafeecb8b434f996ed) to use channels.

Reply:
Added [it](https://github.com/instill-ai/pipeline-backend/pull/819/commits/1d23911395865ae5abe5cde048028315a5b5db8a) back.